internal/planner/planner.go

```diff
@@ -1580,10 +1579,16 @@ func (p *Planner) planRefData(virtual *ruletrie, base *baseptr, ref ast.Ref, ind
 		}); err != nil {
 			return err
 		}
+
+		p.ltarget = base.local
+		return p.planRefDataBaseScan(base.path, ref, index, lexclude, iter)
 	}
 
+	// // nothing to exclude
 	p.ltarget = base.local
-	return p.planRefDataBaseScan(base.path, ref, index, lexclude, iter)
+	return p.planRefRec(ref, 1, func() error {
+		return p.planRefData(nil, &baseptr{local: p.ltarget.(ir.Local)}, ref, index+1, iter)
+	})
 }
```

Comment:
☝️ Took a detour. I think this path is less worrisome, it's basically what happened before minus the (empty) scan.

There's still a bug in how the call_indirect optimization is handled there. It never consults `data` when the optimization has been triggered.

Reply:
https://github.com/open-policy-agent/opa/blob/22bb8fc36b5b0817a9ad274c68fd189853073023/internal/planner/planner.go#L1485
This is what I've been referring to. Slightly adapting the example above, 

```rego
package foo

p {
   k := "q"
   not data.bar[k]
}
```
```rego
package bar

p = 7
```
with data `{"bar": {"q": 10 }}`,
the result of `data.foo.p = x` will be `x: true`.

Without then negation, a similar example can be constructed....